pwndbg/chain.py

```diff
@@ -109,7 +119,9 @@ def format(value, limit=LIMIT, code=True, offset=0, hard_stop=None, hard_end=0,
         hard_stop(int): Value to stop on
         hard_end: Value to append when hard_stop is reached: null, value of hard stop, a string.
         safe_linking(bool): whether this chain use safe-linking
-
+        enhance_can_dereference(bool): whether 'enhance' is allowed to dereference the value it is enhancing
+            Only takes effect when passing a list with a single value, and is used when it may be a pointer that is not meant to be dereferenced
```

Comment:
Why this 'if' that we will dereference when multiple values are passed?

Reply:
I added this parameter as a result of an edge case that came up when making telescoping addresses for annotations. 

As it stood, the `format()` function had an assumption that the list of values you pass as a parameter (`value`) contained a list of pre-dereferenced pointers, where the last value in the list was not a pointer. Something like [0x7fffffffe1e0, 0x7fffffffe2d0, 0x7fffffffe2d8, 0x38]. The code therefore assumes that the second to last value, `chain[-2]`, is a pointer to a non-pointer. This gets passed to enhance, which will dereference it to get an enchanced value. In whole, this was a valid assumption, because we only ever passed the result of `get()` (the function above in the same file) which would always fully dereference a chain of pointers.

However, with doing these annotations, there are cases where we cannot safely dereference a list of pointers that we pass into format(). For example, take the case that we are not emulating and are enhancing an instruction a couple instructions ahead of the PC. If the instruction reads from writable memory, we cannot safely dereference and read the value in memory as it may have been modified by previous instructions. In this case, we pass a single pointer in a list to format, [0x7fffffffe1e0], but we need to tell it that is not safe to dereference and enhance, because enhance will just read from current host memory, and will possibly read an incorrect value (something that has been edited by the time the instruction in questions actually will execute). 

Now that you bring it up, however, it is possible that even with multiple values we could still not want to dereference. I'm imagining the case of `[pointer-to-readonly-memory, pointer-to-writeable-memory]`, which could come up when not-emulating and enhancing future instructions. In this case, we have multiple values in the list, but still don't want to dereference the last value. This made me catch this edge case bug here:

https://github.com/OBarronCS/pwndbg/blob/f110238be083da0f5e779740271c3e3db8bea2d5/pwndbg/disasm/arch.py#L425-L430

This right here handles the case that we are enhancing future instructions without emulation, and we are just using `pwndbg.chain.get`, but we only check if the first pointer is readonly, not subsequent ones. I'll get a fix to this, just manually dereference until we get to a pointer to writable memory.

I'm also considering, maybe instead of making the enhance_can_dereference a parameter, we determine it inside of the format function by seeing if the last value corresponds to a page?

Reply:
This case happens realistically in the following case:

There is an instruction after the current PC, and we are not emulating, like this:

lea    rax, [rip + 0x2e7d]

Where `rip + 0x2e7d` was a pointer to read-only memory, which contains a pointer to writable memory (or a long chain). We should stop dereferencing at the second pointer.

Reply:
Great - latest commits fix this bug - no longer always deference just because multiple values are in the list.

Reply:
Got back to this - newest commit cleans up a lot of this logic and provides an explanation as to why this only applies to lists of length 1

https://github.com/pwndbg/pwndbg/blob/d858f8a75539b74c1e46eef9e75641f9d451ab7a/pwndbg/chain.py#L158-L164